### DISCUSSION TITLE
多人游戏同步问题

### DISCUSSION THREAD

**Discussion by awei-mj** [endink/Mediapipe4u-plugin#67] [Show and tell]
### MediaPipe4U 版本

20230423

### UnrealEngine 版本

5.1.x

### UnrealEngine 类型

源码编译

### 问题类型

运行错误

### 错误详细描述

打算用专用服务器实现多客户端进行客户端与人物模型一一对应的动作捕捉，但如果把运行时组件放在服务端，只有其中一个客户端能控制全部的模型，且只能在该客户端上看到动作，其他客户端无法看到动作也无法控制，已将bReplicates设为true；如果放在客户端，log中会报“LogMediaPipe: Error: The mediapipe component could not be found in the current level”的错，摄像头可以打开但模型无动作。
不了解动画蓝图中的solver是如何实现的，不太清楚到底是插件的问题还是我用得不对
已查看#20，未解决问题

**Comment by endink** [comment]
**首先你应该知道：**
Unreal Engine 的自动同步仅同步 Actor 属性，不会同步动画（骨骼数据）、纹理、网格体信息等，你看到的 Character 上的简单动画被同步是因为 Character 做了属性同步，属性变化（速度，方向，跳跃状态机状态等）被同步，然后动画蓝图在客户端“重放”。

---

MediaPipe4U 是靠 mediapipe 数据播放动画而不是简单靠几个属性，如果要完成网络同步，你需要自己开发 C++ 代码，使用 Unreal Engine 的服务器客户端框架来同步 mediapipe 数据或者直接同步骨骼旋转数据，官方有个血条同步的例子可以参考。

---

对于个人来说，这个工作相当复杂，开发量较大，因此暂时无法支持。毕竟，这基本上是一个用爱发电的项目，无法与公司化运作的产品媲美，实在抱歉，短期不会有这个功能的开发计划~

**Comment by endink** [comment]
简单结论就是，MediaPipe4U 现在没有网络同步功能~

**Comment by awei-mj** [comment]
如果我想进行二次开发，能否在此插件基础上拿到mediapipe数据或者骨骼旋转数据？另外，找不到mediapipe组件的问题可能也需要解决一下。

**Comment by endink** [comment]
Mediapipe 数据可以从 holistic 组件获取到，具体看一下文档进阶用法，当你完成网络同步数据后我可以给你实现重放

**Comment by endink** [comment]
至于找不到组件，查找只是简单遍历，可以禁用自动查找，手动连接组件，具体看MediaPipeAnimationInstsnce 头文件，Connect 相关函数

**Comment by awei-mj** [comment]
你好，我在standalone版中已可以拿到数据，但在专用服务器版中推测由于找不到组件故无法得到数据，且只找到了能在动画蓝图中关闭自动连接mediapipe，未找到动画蓝图的手动连接方法，在头文件中设置bAutoreconnect为true也无效；在C++中实现了手动连接并绑定了事件，但委托函数并未运行，与多播委托/事件绑定的多个函数应该不会互相影响的吧？
望解答，谢谢。

**Comment by endink** [comment]
先梳理一下多人逻辑，mediapipe 从摄像头获取数据的逻辑不应该在 Dedicated Server 运行，他们总是在客户端执行然后同步执行结果给 Dedicated Server，手动连接几乎就是为多人设计的，MediaPipe4U 有一个限制，一个场景内只能有一个 mp holistic (mph)组件，同时也只能有一个animation instance 能够连接到 mph 组件，多个动画连接他会竞争数据，帧率下降，多人游戏肯定要关闭自动连接，只有 pawn 可以连接到 mph ，pawn 自己本地执行动画，并将 MP 数据同步到 server，其他非pawn的动画蓝图从网络中获取数据并重放动画，而不是 mph 中（我们说的M4U不支持网络是因为这部分没有开发）。

开发网络功能你只能用 C++， M4U 在 C++ 中可以访问非常低层的数据，会方便许多。


Connect 和 Disconnect 没有用吗，不太可能

**Comment by endink** [comment]
另外，如果是个人开发，那无所谓，我可以随时和你讨论，如果是公司开发，那要慎重，你正在扩展一个没有源码项目，我未来的代码变更可能会影响到你的代码，因为我不知道你们会用到哪些部分代码，不要继承非接口的类，保持自己的代码的相对独立，尽可能少的耦合 M4U 代码

**Comment by awei-mj** [comment]
mph组件是运行在客户端上的这点没有问题，每个客户端上都会有一个本地的mph，Connect我一会再去尝试一下。关于重现我有个问题，客户端可以通过Actor的ROLE判断其是否为本地控制的，如果为ROLE_Autonomous则为由本地playercontroller控制的，如果为ROLE_Simulated则为由其它客户端控制的，能否将mp数据作为character/animinstance的字段复制到服务端和其他客户端，然后根据前述方法判断所有权，并以此实现动画绑定？
另外，我是在进行毕业设计开发

**Comment by endink** [comment]
当然，你就是需要开发这个过程，至于回放这些数据，我可以在未来版本中暴露一个函数给你

**Comment by endink** [comment]
动画蓝图上有三个函数可以控制是否连接 mph:   
- UMediaPipeAnimInstance::ConnectToMediaPipe
- UMediaPipeAnimInstance::IsMediaPipeConnected()
- UMediaPipeAnimInstance::DisconnectFromMediaPipe()

手动连接设置 UMediaPipeAnimInstance::AutoConnectToMediaPipe 为 Disable   
   
bAutoReconnect 是 private 字段，内部使用的，不是你理解的哪个意思，总之你修改头文件是不行的，你只能调用，这是预编译的库，你怎么能修改头文件呢。。。。  你需要自己的 C++ 代码来设置属性或者调用函数

**Comment by awei-mj** [comment]
了解了，谢谢

**Comment by awei-mj** [comment]
如果我想在c++中使用animinstance，需要继承吗，还是直接使用MediapipeAnimInstance就好
处理网络同步得到的mediapipe数据是调用handle函数吗
如果本地数据仍然想使用solver解算应该如何处理？

**Comment by endink** [comment]
不建议用纯 C++ 处理动画蓝图，那样会非常不通用，而且要求你对动画蓝图代码非常熟悉，比如你要了解 Animation Proxy 这些东西。动画还是建议在蓝图中处理，目前没有支持你自己传入 mediapie 数据，但是用 C++ 你可以这样做：

1. 从 MediapipeAnimInstance 继承一个类。
2. 在你需要更新数据的地方调用 基类的 GetContext 函数得到一个 mediapipe 动画上下文（FMediaPipeAnimContext）。
3. FMediaPipeAnimContext 中有一个属性 MediaPipeSource， 更新这个属性（使用 FMediaPipeSource::UpdateXXX 方法）

因为你的需求比较高端，我没有对这块进行很好的测试，但是 MediaPipe4U 本身也是这样更新数据的，所以应该没有什么问题。

PS: 最后，让用户使用你继承出来的动画蓝图类，为什么要继承？你可以访问 protect 中的函数，做一些“高级”的操作。尽可能找我写好的函数更新数据，不要轻易地改变动画蓝图的属性，因为很多属性可能不是线程安全的，我提供的更新函数都保证了线程安全。

**Comment by endink** [comment]
当你继承 MediapipeAnimInstance 以后，你可以简化上述的步骤，直接使用 MediapipeAnimInstance  的 HandleXXX 更新数据，我没记错的话，HandleXX 可以在任意线程调用而无需担心线程安全问题。

**Comment by awei-mj** [comment]
也就是说，我先继承这个类，然后直接使用HandleXXX更新数据，之后在动画蓝图中使用子类，添加solver节点就行了对吗
更新的数据可以来自本地或网络同步，但solver不需要改变

**Comment by endink** [comment]
是的，简单说就是你不用 Holistic 组件提供数据，自己来提供数据

**Comment by awei-mj** [comment]
只有HandlePoseOut和HandleHandOut，没有HandlePositionOut吗？

**Comment by endink** [comment]
Position 不是 mediapipe 数据，只需要提供：

- Pose landmarks
- LeftHand Landmarks
- RightHand Landmarks
- PoseWordLandmarks

表情数据你可以自行同步动画曲线，然后用，pose asset 回放，表情无法这样同步，有映射问题。

**Comment by endink** [comment]
当然，还有 Image source Info 也是必需的

**Comment by awei-mj** [comment]
代码写完了，但是好像一开始的绑定监听事件就没绑上
本地控制的holistic组件(服务端上没有，每个客户端上有一个)，也找到了该组件，但绑定的函数没有执行，伪代码大概是这样的，能看到第一个log但看不到第二个
```c++
if(HolisticComponent){
    UE_LOG(...);
    HolisticComponent->xxxTrigger().AddUObject(this, &xxCharacter::Handlexxx);
}

//绑定的监听函数里
UE_LOG(...);
//处理数据
```
本来怀疑是因为客户端的Character只是副本，又用本地的Actor试了一下还是绑不上

**Comment by endink** [comment]
请在 tick 中持续检查你的 component 有效性和 Holistic 有效性，Component 在 Edito中有不一般的生命周期，这不是一句话能说清楚的，Editor 不是你想象的那样简单，我想是组件重新加载了

**Comment by awei-mj** [comment]
holistic应该一直都是有效的，但是问题依旧
另外每次重启UE编辑器动画蓝图中的MP节点都会消失

**Comment by endink** [comment]
你需要进一步调试委托链甚至UE源码确定哪个环节发生问题，现有信息我无法再提供建议
